pkg/ottl/contexts/internal/ctxutil/slice.go

```diff
@@ -5,61 +5,258 @@ package ctxutil // import "github.com/open-telemetry/opentelemetry-collector-con
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
+	"golang.org/x/exp/constraints"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
 )
 
-func GetSliceValue[K any](ctx context.Context, tCtx K, s pcommon.Slice, keys []ottl.Key[K]) (any, error) {
+const (
+	typeNotIndexableError = "type %T does not support indexing"
+)
+
+var (
+	errMissingSetKey = errors.New("cannot set slice value without key")
+	errMissingGetKey = errors.New("cannot get slice value without key")
+)
+
+func getSliceIndexFromKeys[K any](ctx context.Context, tCtx K, sliceLen int, keys []ottl.Key[K]) (int, error) {
 	if len(keys) == 0 {
-		return nil, fmt.Errorf("cannot get slice value without key")
+		return 0, fmt.Errorf("cannot get/set slice index without a key")
 	}
 
 	i, err := keys[0].Int(ctx, tCtx)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 	if i == nil {
 		resInt, err := FetchValueFromExpression[K, int64](ctx, tCtx, keys[0])
 		if err != nil {
-			return nil, fmt.Errorf("unable to resolve an integer index in slice: %w", err)
+			return 0, fmt.Errorf("unable to resolve an integer index in slice: %w", err)
 		}
 		i = resInt
 	}
 
 	idx := int(*i)
 
-	if idx < 0 || idx >= s.Len() {
-		return nil, fmt.Errorf("index %d out of bounds", idx)
+	if idx < 0 || idx >= sliceLen {
+		return 0, fmt.Errorf("index %d out of bounds", idx)
+	}
+
+	return idx, nil
+}
+
+func GetSliceValue[K any](ctx context.Context, tCtx K, s pcommon.Slice, keys []ottl.Key[K]) (any, error) {
+	if len(keys) == 0 {
+		return 0, errMissingGetKey
+	}
```

Comment:
Do we want to just lean on this same check in `getSliceIndexFromKeys`?

Reply:
The idea here was provide a more specific message with the world `get` or `set` on it, so statements like `set(foo, slice_path[])` would generate a more clear message `cannot **get** slice value without key`.

Another option is to move all validations into the `getSliceIndexFromKeys` (used for getters and setters), and add an extra argument to proper format the error message. WDYT?

Reply:
Yes, that sounds good to me.